onnxscript/ir/passes/common/shape_inference.py

```diff
@@ -0,0 +1,98 @@
+
+    def call(self, model: ir.Model) -> ir.passes.PassResult:
+        # Store the original initializer values so they can be restored
+        initializer_values = tuple(model.graph.initializers.values())
+        tensors = {v.name: v.const_value for v in initializer_values}
+        original_inputs_len = len(model.graph.inputs)
+        initializer_names = {v.name for v in initializer_values}
+
+        # Turn the initializers into inputs and clear the initializers
+        # to limit the model size
+        for initializer in initializer_values:
+            # Make sure the initializer has its shape/type set
+            assert initializer.const_value is not None
+            if initializer.shape is None:
+                initializer.shape = initializer.const_value.shape  # type: ignore[assignment
+            if initializer.dtype is None:
+                initializer.dtype = initializer.const_value.dtype
+            if initializer not in model.graph.inputs:
+                model.graph.inputs.append(initializer)
+            initializer.const_value = None
+        model.graph.initializers.clear()
+
+        # Perform shape inference
+        try:
+            proto = ir.serde.serialize_model(model)
+            proto = onnx.shape_inference.infer_shapes(
+                proto,
+                check_type=self.check_type,
+                strict_mode=self.strict_mode,
+                data_prop=self.data_prop,
+            )
+            inferred_model = ir.serde.deserialize_model(proto)
+        except Exception:  # pylint: disable=broad-exception-caught
+            logger.warning("Shape inference failed. The model is not modified", exc_info=True)
+            return ir.passes.PassResult(model, modified=False)
+        finally:
+            # Restore the original initializer values so the model is unchanged
+            for initializer in initializer_values:
+                if initializer.name in initializer_names:
+                    initializer.const_value = tensors[initializer.name]
+                    model.graph.register_initializer(initializer)
+
+            # Restore the original inputs
+            inputs = model.graph.inputs[:original_inputs_len]
+            model.graph.inputs.clear()
+            model.graph.inputs.extend(inputs)
+
+        # Add the original initializer tensors to the new (inferred) model
+        for new_input in inferred_model.graph.inputs:
+            # Assign the tensors back to the initializers
+            if new_input.name in initializer_names:
+                new_input.const_value = tensors[new_input.name]
+                inferred_model.graph.register_initializer(new_input)
+
+        # Remove the inputs that were added
+        new_inputs = inferred_model.graph.inputs[:original_inputs_len]
+        inferred_model.graph.inputs.clear()
+        inferred_model.graph.inputs.extend(new_inputs)
+        # Even though modified, we know the pass will not change the model if we ran it again.
```

Comment:
Seems like we need some thoughts/documentation on what `modified` means ... looks like it could have several different meanings (potentially distinct). Another dimension here is the `in_place` ... so, if in_place is false, the incoming model is guaranteed not to be modified? so, this `modified` field (when in_place=False) could refer to (a) is outgoing model same as incoming model, or (b) will running this pass again on outgoing model produce a different result.

Seems to me (a) is better? Also: we can't guarantee (b) without making some assumptions about ONNX's shape-inference implementation details.

Reply:
It's intended to be used as (b) (i.e. when to early stop if we run iterate over passes a few times) but since there is no way we know (b) we can only use (a) when there is not enough information I think?

Reply:
Done. Properly set modified field